### PR TITLE
LPS-115484 As a developer I can declare theme's customizable CSS variables

### PR DESCRIPTION
**Pull request by wincent**
This was sent to me at https://github.com/wincent/liferay-portal/pull/357 — tests failed there, so I am going to take that as an opportunity to relocate this here instead and re-run them. Note the dependency on a prior pull (https://github.com/wincent/liferay-portal/pull/354) — the last three commits are the new ones, based on top of the other pull.

cc @izaera 

Original message follows.

---

This is the continuation of https://github.com/wincent/liferay-portal/pull/354.

This is a partial PR based on the original PoC https://github.com/jbalsas/liferay-portal/pull/2211. It adds a service to retrieve CSS variable declarations inside themes' WAR files. These declarations will then be used to create customization forms in future PRs. Then, those forms will save configuration values that will be retrieved by a `ScopedCSSVariablesProvider` to inject them in the rendered HTML.

There's nothing to test manually in this PR as it is not yet possible to do anything from the GUI.

### CODE REVIEW

**Comment by wincent** [comment]
ci:test:sf

**Comment by wincent** [comment]
ci:test

**Comment by wincent** [comment]
Same test failures here as in https://github.com/liferay-frontend/liferay-portal/pull/21 so assigning to you @izaera to check those out.

**Comment by jbalsas** [comment]
Hey @izaera, do we need LPS-115485 commits in here? If we do, we'll need to wait for those to get merged to safely push this... if they aren't, could you take them out so we can work on this independently?

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+	public void modifiedService(
+		ServiceReference<ServletContext> serviceReference,
+		Map<String, CSSVariableDescription> cssVariablesDescriptions) {
+
+		ServletContext servletContext = _bundleContext.getService(
+			serviceReference);
+
+		try {
+			InputStream is = servletContext.getResourceAsStream(
+				"/WEB-INF/css-variables.json");
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:78]: Let's make this configurable and give it a nice default, just in case...

Comment (jbalsas) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:78]: Let's call the default `tokens.json` or something like that so we don't tie this to a particular implementation.

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+			labelsMap.put(StringPool.BLANK, label);
+		}
+
+		return labelsMap;
+	}
+
+	private Map<String, CSSVariableDescription> _parse(String json)
+		throws JSONException {
+
+		JSONObject jsonObject = _jsonFactory.createJSONObject(json);
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:160]: Could you please define the JSON object model somewhere? Feels like an important part of this that is not reflected anywhere.

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java (hunk 1 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:78]: You mean the file name? Why would we want to make it configurable? :thinking:

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java (hunk 2 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:160]: Maybe we can create a JSON schema file, but I'm not sure where to put it (portal, Themes Tookit, ...) :thinking:

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java (hunk 1 above)
Comment (jbalsas) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:78]: Because we often shoot ourselves in the foot and change the name later or do some other weird stuff.

Just define a `tokensPath` or something like that and make `tokens.json` the default in case nothing is passed. We can then generate dev tooling for this without hardcoding "css-variables.json"  everywhere in different projects and forcing users down that name.

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java (hunk 2 above)
Comment (jbalsas) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:160]: I see it in https://github.com/liferay-frontend/liferay-portal/pull/17/commits/573a9b32b9d010f06cc8410b8ecdc0a733ccc293#diff-890f20d666168a89aec122633df87dcc, so I'll comment there.

File: modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+	"variables": {
+		"background": {
+			"label": "Background Color",
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json:4]: So... themes can't have ResourceBundles, do they?

I'm wondering if we could make this go through `LanguageUtil.get()` and avoid having to support localization in a json... this feels like a weird thing to build in here, though granted, we've done this in the past...

Thinking our efforts might be more valuable in figuring out how to provide a `ResourceBundle` for this than adding multilang support in here...

File: modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json [hunk 4]
```diff
@@ -0,0 +1,14 @@
+{
+	"variables": {
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json:2]: I'm also thinking how to process this without equating `object` keys to `css variables`...

```json
{
    "tokens": [
        {
            cssVariable: "background",
            label: "background-color",
            type: "color"
        }, {
            cssVariable: "font",
            label: "font",
            type: "text"
        }
    ]
}
```

Wouldn't this give us some additional room to expand the API in the future without compromising the design? If we rename this to `tokens.json` then it makes it more general and it doesn't feel so constrained to specific `css-variables`.

What do you think?

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java (hunk 1 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:78]: I agree with the renaming part :sweat_smile: . 

The problem with making it configurable is that once a tool/portal version starts using a certain location making configuration changes can only lead to disaster because then you have to care about tooling using old names, new names, and how to setup the portal. 

Also, you may need to deploy WAR files using the old location for the file and WARs using the new one, but you may only configure one value in the portal, unless we add support for multiple places.

All other configuration files (MANIFEST.MF, OSGi metatype, package.json and so on) are placed in well-known locations for that reason.

I'll extract it to a constant/property, anyway, though I have used constants in the past for the very same reasons you are stating and they were removed in favour of hard coded strings in revisions...

File: modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json (hunk 3 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json:4]: No, they can't because they are WAR files. I will investigate how easy it is to instantiate a ResourceBundle on the fly or if there's something available...

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java (hunk 1 above)
Comment (jbalsas) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:78]: By configurable I mean in the `package.json`, not as a Portal Setting

File: modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json (hunk 4 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json:2]: I also thought about making this json file more general, but didn't know how to name it or organize it. 

Maybe we can be ambitious and name it something very generic like `theme.json` so that we can use the same file infrastructure in the future to add any information we think is relevant. 

Then we would name the section something like `css-variables` to leave room for any other data we may add in the future.

Does that make sense?

Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json:2]: BTW, I don't like `theme.json`, but I cannot think of any good name now. However we can defer that a bit until we forward this pull.

File: modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json (hunk 3 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json:4]: There's no way to add language keys to WAR files. So we may:

1. Leave localization in this JSON
2. Move the localization to properties files and create some sort of service that reads them from WAR files and instantiates `ResourceBundle`s
3. Migrate themes to JAR files
4. Don't localize

4 doesn't seem an option.

3 we already investigated it and was way more difficult than we thought. It would also imply migrating the toolkit.

2 is a good thing if we know for sure that we won't do 3 in the near future. The idea is that WAR files are kind of deprecated, so there's no point in making the migration more difficult by supporting newer things. However, WAR files will always work, and themes don't need Java classes, so I don't see it so bad to leave themes as WARs forever.

What do you think?

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java (hunk 1 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:78]: Just to confirm: you mean adding a known key in the `liferayTheme` section of the `package.json` to specify where this file is stored inside the WAR?

File: modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json (hunk 3 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json:4]: BTW, for 2 we would use the same techniques we use in the JS Portlet Extender (instantiate an OSGi service on the fly configured by a certain file in the bundler/WAR).

**Comment by izaera** [comment]
@jbalsas 

> Hey @izaera, do we need LPS-115485 commits in here?

Yes, but only because they create the project folders with the gradle and bnd files. But other than that, nothing else is shared, so we may work on this PR independently from the other (that's why I made two PRs in first place).

It's only that we cannot forward this until the other gets merged which, on the other hand, wouldn't make too much sense.

File: modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json (hunk 4 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/css-variables.json:2]: Thinking about this after [this reply](https://github.com/liferay-frontend/liferay-portal/pull/17#discussion_r443542647) my suggestion may not apply any more, as we may declare as many new files as we want in the package.json, thus we don't need a single well-known file any more, nor create sections inside it.

Regarding your suggestion (using an array of objects instead of an object of objects) it's OK. I find it a bit more verbose, but it is true that it's more flexible.

And for the name of the file, it wouldn't be very important, because we let users configure it in the package.json file. 

We would still need to decide if we want it to be only for css variables or leave room for something else. This is something I'm not 100% sure of, but I would say that given that we don't have any other use in mind now (at least me) it would be better to make it clear that the file is for css-variables, not a vague term like `tokens`. 

Because if we name it tokens, we are making it more difficult to understand for the final user, adding another term which is not yet defined. And it's true that we may find something else that fits into the `tokens` idea in the future, but it's also very possible that we end up mixing miscellaneous unrelated things in the same file just because they are "tokens".

So unless you have a clear idea in your mind of what we will use the term `tokens` for in the future, I would consider it a premature generalization.

**Comment by izaera** [comment]
@jbalsas I have force-pushed (because I needed to rebase) the requested changes. Do you mind having a look in case something else needs to be polished? Thx.

The real changes start at commit 2066d40 (i.e.: only the last three are new) with the last one being the more important. The rest remain the same, as in the previous PR.

I have implemented the change from `css-variables.json` to `tokens.json` even though we didn't get to an agreement in [this thread](https://github.com/liferay-frontend/liferay-portal/pull/17#pullrequestreview-434831256)...

**Comment by izaera** [comment]
ci:test:sf

**Comment by izaera** [comment]
ci:test:relevant

File: modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/liferay-look-and-feel.xml [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0"?>
+<!DOCTYPE look-and-feel PUBLIC "-//Liferay//DTD Look and Feel 7.0.0//EN" "http://www.liferay.com/dtd/liferay-look-and-feel_7_0_0.dtd">
+
+<look-and-feel>
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/liferay-look-and-feel.xml:4]: Does this file need to be so "complete"?

Since it's for a mock... could we just leave here the least amount of necessary information? I got pretty scared for a while thinking we had to add a whole theme or something for this...

Also, it'd help to make this explicit if we did something like `theme id="testTheme" name="Test Theme"`.

just-sayin

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeBundleInspector.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+import java.util.Map;
+
+import org.osgi.framework.Bundle;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+public class ThemeBundleInspector {
+
+	public static boolean isTheme(Bundle bundle) {
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeBundleInspector.java:46]: There seems to exist `pluginAutoDeployListenerHelper.isThemePlugin()`... 

```java
public boolean isThemePlugin() throws AutoDeployException {
	if (isMatchingFile("WEB-INF/liferay-look-and-feel.xml", false)) {
		return true;
	}

	Matcher matcher = _themePluginPattern.matcher(_file.getName());

	if (matcher.find() &&
		isMatchingFile(
			"WEB-INF/liferay-plugin-package.properties", false)) {

		return true;
	}

	return false;
}
```

Could we use it or at least have the same implementation?

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	@Override
+	public void modifiedService(
+		ServiceReference<ServletContext> serviceReference,
+		AtomicReference<Map<String, CSSVariableDescription>> atomicReference) {
+
+		Bundle bundle = serviceReference.getBundle();
+
+		if (!ThemeBundleInspector.isTheme(bundle)) {
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:64]: Feels like a lot of duplicated `.isTheme` here and in the `ThemeBundleInspector` constructor.

I know it's safer, but the current code path can't throw the exception because you're checking here earlier in a static method to see if the bundle is a theme... so what's the point either in checking here or in throwing there?

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+					themeBundleInspector.getSymbolicName(),
+				jsonException);
+		}
+	}
+
+	@Override
+	public void removedService(
+		ServiceReference<ServletContext> serviceReference,
+		AtomicReference<Map<String, CSSVariableDescription>> atomicReference) {
+	}
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:91]: Why don't we need to do anything here? I'm sure it's fine, but I'd like to be able to follow this.

File: modules/apps/frontend-css/frontend-css-variables-api/src/main/java/com/liferay/frontend/css/variables/CSSVariableDescription.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.css.variables;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+public interface CSSVariableDescription {
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-api/src/main/java/com/liferay/frontend/css/variables/CSSVariableDescription.java:20]: This would be a TokenDescription now...

File: modules/apps/frontend-css/frontend-css-variables-api/src/main/java/com/liferay/frontend/css/variables/CSSVariableDescription.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+	public CSSVariableType getCSSVariableType();
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-api/src/main/java/com/liferay/frontend/css/variables/CSSVariableDescription.java:22]: `TokenType`? :D

File: modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/liferay-look-and-feel.xml (hunk 5 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/test/resources/com/liferay/frontend/css/variables/web/internal/theme/liferay-look-and-feel.xml:4]: OK. Sounds good. :+1:

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java (hunk 7 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:64]: Defensive programming and failing early. Otherwise, we can create a `ThemeBundleInspector` for a non-theme bundle and call the other methods and they may fail or return bogus values randomly.

I also thought about the duplication but couldn't find a better pattern :thinking:. I tried:

1. The static method to check before construction so that you don't need to rely on a exception to do flow control 
2. Only check in constructor: you don't check twice, but the only way to test if a bundle is a theme is to construct and catch exceptions
3. An `isTheme` method in the object. This is cleaner, but then you need to check `isTheme` in every other method to avoid returning invalid values.

Another possibility could be a factory method to check and construct the object so that if the bundle is not a theme it returns null.

I don't know which one I prefer. Also I'm not sure which one will pass further revisions better...

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java (hunk 8 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:91]: :scream: Of course we need to do something: remove the variable descriptions :sweat_smile:

Comment (jbalsas) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java:91]: 😉

**Comment by izaera** [comment]
> I'm trying to wrap up these changes with the ones we already pushed to master... sorry, getting hard to keep track of everything :(
> I'd say we need something that's called css-variables which is the thing we definitely use to print the CSS variables.

I split it up into two pulls for that reason ;-). The first PR is `css-variables` for sure but this one, I'm not sure... 

I would say this is CSS variables too because it only adds a service to discover which CSS variables a theme supports. And we know that they are CSS variables because it's a theme. However, I see place for tokens in a theme (for example inside templates, or predefined content) so I'm not sure what is best:

1. Leave this as CSS variables and implement a new service in the future if we add more tokens
2. Declare everything as a token (CSS variables and what may come next) and rename this whole PR

The only thing that scares about 2 is that tokens may be different things covered under the same abstraction in the future. For example: CSS variables now have a name and type. One of the types is `color` which is OK for a CSS variable but may be totally irrelevant for a content token (for example). You could argue that in that case you use `string` which is OK, but then more types can appear, and more attributes which are not the same for every type of token.

To sum up: I'm OK with naming this tokens if we at least have a clear idea of what other tokens may look like (i.e: if we can define what is a token and what isn't now, in advance).

**Comment by jbalsas** [comment]
Putting this on hold momentarily since TokenDefinition is in a bit of a definition phase itself :)

**Comment by jbalsas** [comment]
Pinging @ealonso, @ruben-pulido and @JorgeFerrer here so we can start this again. This is an initial PR we had before the StyleEditor epic started moving that does a bunch of things that might not be necessary.

While we [discuss the JSON format](https://github.com/jbalsas/liferay-portal/pull/2218) I'm looking for **input** as to what to do in regards:
- How to name this thing? `tokens.json`? `design_system.json`? `theme.json`?
- This PR offered a way to retrieve that definition of a theme to be consumed. The question here is... should the definition be returned as:
    - Plain JSON Object
    - A specific Java model

**Comment by JorgeFerrer** [comment]
The name of the file must reflect as accurately as possible what it is, not what it's used for.
In this case, that means that it should have the word "token" in it, since what the file does is define tokens. An alternative term could be "css-variables" since that's at the end what a token is.

Since this is a definition of what the available tokens are, I believe that token-definition.json is a good name.

Thoughts?

I don't have a clear opinion on the JSON vs Java question right now, so I'd prefer other people to chime in on that one.

**Comment by jbalsas** [comment]
Hey @izaera, I'm going to summarize my understanding of what we have and what I think we could build next to try to unblock this.

Please, correct me as needed! 🤗 

### frontend-css-variables modules
We currently have 2 `frontend-css-variables-*` modules (`api` and `web`) that are the ones responsible for registering and rendering values for scoped css variables. 

### theme token definition
Our next step would be to provide an interface to:
- Enhance Base Themes with a `token-definition.json` 
- Provide APIs to:
    - Retrieve a TokenDefinitionJSON given a `themeId`
    - ~Retrieve a List of [themeID, tokenDefinition].~ Not sure if this one's needed right now, though... so maybe better keep it out for now

Waiting on @ealonso and @ruben-pulido to answer https://github.com/liferay-frontend/liferay-portal/pull/17#issuecomment-654130309, I think we could simply return the `JSON` or even `String` here. We can always enhance the API to model the `TokenDefinition`, but maybe it's too son for it. What do you think?

Based on this, I'd suggest:
- Create a new module to enhance and extend base themes (wars) with `TokenDefinition` and provide the mentioned API. 

@JorgeFerrer, does this seem aligned with Echo's roadmap? (can't ping Tarik since he still doesn't have a GitHub alias... but sending him a message on Slack)

**Comment by ealonso** [comment]
> Waiting on @ealonso and @ruben-pulido to answer [#17 (comment)](https://github.com/liferay-frontend/liferay-portal/pull/17#issuecomment-654130309), I think we could simply return the `JSON` or even `String` here. We can always enhance the API to model the `TokenDefinition`, but maybe it's too son for it. What do you think?

A String should be enough at this moment.


> @ealonso, @ruben-pulido, do you think this should be enough to implement the StyleBook creation or would you require something else?

What do you mean with, **this should be enough**?

**Comment by jbalsas** [comment]
> What do you mean with, this should be enough?

So, once we do this we'll have:
- A way to get a `TokenDefinition` for a theme. This should allow you to generate UI for those tokens and store values where you want.
- A way to provide token values as `ScopedCssVariable` so the values in a StyleBook apply to a given scope. This should allow for stored StyleBooks to have visible changes

Do you think there's any additional Infrastructure or Component you'd like us to provide or work on?

For example, we discussed having an autogenerated UI which was discarded. Another thing that we could explore is providing a unified way of "previewing" changes. This is mostly tied to the UI, and likely some simple JS based on our past experience, but we can flesh it out more if needed.

**Comment by izaera** [comment]
> Please, correct me as needed!

LGTM :-)

**Comment by izaera** [comment]
> An alternative term could be "css-variables" since that's at the end what a token is.

If a token is really a CSS variable, I would call it CSS variable, since that way you don't need two terms. 

However, I think token is higher level than CSS variable, since you may have an UI to define tokens that will be injected as CSS variables in themes, but could be injected as some other thing (for example, an inline style) in other place of the portal. Theoretically you could even use a color token to colorize images/logos.

Said that, the JSON file that goes inside a theme file, AFAIK, can only be considered as a description of CSS variables as they will only be used inside CSS rules (that is: the JSON file describes low level CSS variables, not high level UI tokens). Unless we plan to use them in templates, or other content inside theme bundles...

**Comment by jbalsas** [comment]
> Said that, the JSON file that goes inside a theme file, AFAIK, can only be considered as a description of CSS variables as they will only be used inside CSS rules (that is: the JSON file describes low level CSS variables, not high level UI tokens). Unless we plan to use them in templates, or other content inside theme bundles...

That JSON must have a way to map tokens with CSSVariables, but in no way is a description of CSS Variables.

Now that you mention it... one way to simplify what we're discussing over at https://github.com/jbalsas/liferay-portal/pull/2218 would be to split into 2 files... `token-definition.json` and `css-var-mappings.json`... 🤔

**Comment by izaera** [comment]
> would be to split into 2 files... `token-definition.json` and `css-var-mappings.json`

But wouldn't that be a 1-1 mapping?

To elaborate more my view: my assumptions are that a theme bundle is deployed and, because it is a theme, it defines the CSS variables it lets override.

Then, some OSGi bundle in the portal reads those CSS variable definitions, interprets them as tokens, and shows a UI to the user to define those tokens (and possibly some tokens coming from other non-theme places).

Then, the OSGi bundle that showed the UI to define those tokens takes care of registering a ScopedCSSVariablesProvider to inject the values for the theme (and possibly other services to render the non-CSS tokens).

Does that makes sense?

I mean, why letting the theme know anything about tokens? What is the advantage?

**Comment by jbalsas** [comment]
> Does that makes sense?

It kinda does, yeah... but I think it's limiting in what we want to do for the future

> I mean, why letting the theme know anything about tokens? What is the advantage?

I think @JorgeFerrer has a pyramid with this and what we want to accomplish 😂 . If you check Tarik's presentation, it was outlined there, and some things we'd need for the future might or might not be implemented through CSS Variables.

I think it's rather unfortunate that right now this is our only possible way forward. We shouldn't bank all our strategy (naming, APIs, roadmap) in this, though.

**Comment by izaera** [comment]
> We shouldn't bank all our strategy (naming, APIs, roadmap) in this, though.

That's why I was proposing to call it CSS variables (only inside themes) which is the only thing we have clear right now ;-). 

However, given that we are only going to return a JSON string I will call it tokens for now and we'll see what happens in the future :sweat_smile:

**Comment by ealonso** [comment]
> So, once we do this we'll have:
> 
> * A way to get a `TokenDefinition` for a theme. This should allow you to generate UI for those tokens and store values where you want.
> * A way to provide token values as `ScopedCssVariable` so the values in a StyleBook apply to a given scope. This should allow for stored StyleBooks to produce visible effects on pages
> 
> Do you think there's any additional Infrastructure or Component you'd like us to provide or work on?

I think it is good enough for the moment, once we start working with it, we can give you more feedback

**Comment by jbalsas** [comment]
> I think it is good enough for the moment, once we start working with it, we can give you more feedback

Great, thanks!!

@izaera, going to close this one since the solution will be pretty different... please link to this when you open a new one? :pray:

File: modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeBundleInspector.java (hunk 6 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeBundleInspector.java:46]: Yes.

**Comment by izaera** [comment]
This is hopefully the final PR -> https://github.com/brianchandotcom/liferay-portal/pull/91176

Linking for the record.
